### PR TITLE
assert_equal_hash makes large hash comparisons easier to understand

### PR DESCRIPTION
**Pull request by martinemde**
Two differently sorted hashes are equal in the eyes of `assert_equal`, but their diff output can be a huge mess due to the order of hashes (which does not affect equality).

This PR adds a method that sorts any hashes in expected and actual before comparing them in order to improve the readability of the output when they are not equal.

Also converts any multi-line hash comparisons in tests to use the new method.

### CODE REVIEW

File: test/test_helper.rb [hunk 1]
```diff
@@ -144,6 +144,20 @@ def assert_event(tag, expected_additional, actual)
     assert_equal actual.additional_type.new(user_agent_info:, **expected_additional), actual.additional
   end
 
+  # Asserts that both are the same nested hash, using equality checks unless
+  # the value is a hash, in which case it will recursively check the nested hash
+  # The purpose of this assertion is mostly to provide a more useful error when
+  # a nested hash contains different elements instead of overwhelming the user
+  # with a giant nested hash diff that is mostly sorting problems.
```
Comment (simi) [inline test/test_helper.rb:151]: :thinking: isn't Hash not sorted by design?

Comment (martinemde) [reply test/test_helper.rb:151]: Yes, the problem is when you compare the hashes. If they are different by one key you get a huge diff of all the keys that are out of order between the expected and actual with the 1 key that is wrong hidden somewhere, not even necessarily on lines that are close together.

**Comment by martinemde** [comment]
Example failure before:

```diff
  1) Failure:
Avo::UsersSystemTest#test_change_user_email [test/system/avo/users_test.rb:522]:
Expected equal elements in expected.
--- expected
+++ actual
@@ -1,16 +1,18 @@
 {"records"=>
   {"gid://gemcutter/User/78070"=>
     {"changes"=>
-      {"updated_at"=>["2024-12-03T20:08:36.805Z", "2024-12-03T20:08:37.715Z"],
-       "email"=>["user1@example.com", "gem-maintainer-001@example.com"],
+      {"email"=>["user1@example.com", "gem-maintainer-001@example.com"],
+       "token_expires_at"=>
+        ["2024-12-03T23:08:36.805Z", "2024-12-03T23:08:37.711Z"],
        "email_confirmed"=>[true, false],
        "confirmation_token"=>
         ["5295949011b084b54aa2ee4dbace5e858973bc9f1e1946d9",
          "80cbfcc7c7ffb4b4186aa6249c638c5390ca1e8385ab2903"],
-       "token_expires_at"=>
-        ["2024-12-03T23:08:36.805Z", "2024-12-03T23:08:37.711Z"]},
+       "updated_at"=>["2024-12-03T20:08:36.805Z", "2024-12-03T20:08:37.715Z"]},
      "unchanged"=>
-      {"id"=>78070,
+      {"unconfirmed_email"=>nil,
+       "mfa_level"=>"disabled",
+       "id"=>78070,
        "encrypted_password"=>
         "$2a$04$OJFcViJPWq1Ru1yFi5KigeBs./Fg0HdOMF8EbZxC9GD5mbWFzp7qG",
        "salt"=>nil,
@@ -22,9 +24,7 @@
        "handle"=>"handle1",
        "hide_email"=>true,
        "twitter_username"=>nil,
-       "unconfirmed_email"=>nil,
        "remember_token_expires_at"=>nil,
-       "mfa_level"=>"disabled",
        "mail_fails"=>0,
        "blocked_email"=>nil,
        "webauthn_id"=>
@@ -46,41 +46,14 @@
         [nil,
          {"email"=>nil,
           "user_agent_info"=>
-           {"os"=>"Mac OS X",
+           {"installer"=>"Browser",
             "device"=>"Mac",
-            "system"=>nil,
-            "installer"=>"Browser",
+            "os"=>"Mac OS X",
             "user_agent"=>"HeadlessChrome",
-            "implementation"=>nil}}],
+            "implementation"=>nil,
+            "system"=>nil}}],
        "created_at"=>[nil, "2024-12-03T20:08:37.729Z"],
        "updated_at"=>[nil, "2024-12-03T20:08:37.729Z"]},
-     "unchanged"=>{}},
-   "gid://gemcutter/Events::UserEvent/97495"=>
-    {"changes"=>
-      {"id"=>[nil, 97495],
-       "tag"=>[nil, "user:email:sent"],
-       "trace_id"=>[nil, "0"],
-       "user_id"=>[nil, 78070],
-       "ip_address_id"=>[nil, 6797],
-       "geoip_info_id"=>[nil, 811],
-       "additional"=>
-        [nil,
-         {"to"=>"gem-maintainer-001@example.com",
-          "from"=>"no-reply@mailer.rubygems.org",
-          "action"=>"email_confirmation",
-          "mailer"=>"mailer",
-          "subject"=>"Please confirm your email address with RubyGems.org",
-          "message_id"=>
-           "674f6545e36c0_8ab110d2430849@Martins-M3-Max.local.mail",
-          "user_agent_info"=>
-           {"os"=>"Mac OS X",
-            "device"=>"Mac",
-            "system"=>nil,
-            "installer"=>"Browser",
-            "user_agent"=>"HeadlessChrome",
-            "implementation"=>nil}}],
-       "created_at"=>[nil, "2024-12-03T20:08:37.948Z"],
-       "updated_at"=>[nil, "2024-12-03T20:08:37.948Z"]},
      "unchanged"=>{}}},
  "fields"=>{"from_email"=>"gem-maintainer-001@example.com"},
  "arguments"=>{},
```

With this change:

```diff
  1) Failure:
Avo::UsersSystemTest#test_change_user_email [test/system/avo/users_test.rb:522]:
Expected equal elements in expected.
--- expected
+++ actual
@@ -23,33 +23,6 @@
        "updated_at"=>[nil, "2024-12-03T20:18:59.654Z"],
        "user_id"=>[nil, 78103]},
      "unchanged"=>{}},
-   "gid://gemcutter/Events::UserEvent/97561"=>
-    {"changes"=>
-      {"additional"=>
-        [nil,
-         {"action"=>"email_confirmation",
-          "from"=>"no-reply@mailer.rubygems.org",
-          "mailer"=>"mailer",
-          "message_id"=>
-           "674f67b3ce706_e5a48674491be@Martins-M3-Max.local.mail",
-          "subject"=>"Please confirm your email address with RubyGems.org",
-          "to"=>"gem-maintainer-001@example.com",
-          "user_agent_info"=>
-           {"device"=>"Mac",
-            "implementation"=>nil,
-            "installer"=>"Browser",
-            "os"=>"Mac OS X",
-            "system"=>nil,
-            "user_agent"=>"HeadlessChrome"}}],
-       "created_at"=>[nil, "2024-12-03T20:18:59.862Z"],
-       "geoip_info_id"=>[nil, 832],
-       "id"=>[nil, 97561],
-       "ip_address_id"=>[nil, 6818],
-       "tag"=>[nil, "user:email:sent"],
-       "trace_id"=>[nil, "0"],
-       "updated_at"=>[nil, "2024-12-03T20:18:59.862Z"],
-       "user_id"=>[nil, 78103]},
-     "unchanged"=>{}},
    "gid://gemcutter/User/78103"=>
     {"changes"=>
       {"confirmation_token"=>
```

File: test/test_helper.rb (hunk 1 above)
Comment (martinemde) [reply test/test_helper.rb:151]: I went ahead and changed the assertion to sort everything, then compare as normal. It dups everything but I think that's fine considering how much easier the output is to read. See comment below.

**Comment by martinemde** [comment]
For reference, this seems [not desired upstream](https://github.com/minitest/minitest/issues/1016).

**Comment by martinemde** [comment]
ok, the changes needed a bunch of indentation fixes, but nothing changed except getting the hashes into the assertion in a clean way.

**Review by indirect — APPROVED** [review]
nice, thanks for improving this

**Pull request merged**